src/grammar/grammar.ts

```diff
@@ -542,16 +564,35 @@ export class AttributedScopeStack {
 	public getScopeNames(): string[] {
 		return this.scopePath.getSegments();
 	}
+
+	public getExtensionIfDefined(base: AttributedScopeStack | null): AttributedScopeStackFrame[] | undefined {
+		const result: AttributedScopeStackFrame[] = [];
+		let self: AttributedScopeStack | null = this;
+
+		while (self && self !== base) {
+			result.push({
+				encodedTokenAttributes: self.tokenAttributes,
+				scopeNames: self.scopePath.getExtensionIfDefined(self.parent?.scopePath ?? null)!,
+			});
+			self = self.parent;
+		}
+		return self === base ? result.reverse() : undefined;
+	}
+}
+
+interface AttributedScopeStackFrame {
+	encodedTokenAttributes: number;
+	scopeNames: string[];
 }
 
 /**
  * Represents a "pushed" state on the stack (as a linked list element).
  */
-export class StateStack implements StackElementDef {
+export class StateStackImpl implements StateStack {
 	_stackElementBrand: void = undefined;
 
 	// TODO remove me
-	public static NULL = new StateStack(
+	public static NULL = new StateStackImpl(
```

Comment:
We could remove `!` from a few lines below here since now these fields are defined as nullable.